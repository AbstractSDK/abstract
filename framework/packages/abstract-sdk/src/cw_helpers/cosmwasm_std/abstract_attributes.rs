use abstract_std::ABSTRACT_EVENT_TYPE;
use cosmwasm_std::{Attribute, Event, Response};

/// Trait for adding `abstract` attributes to something
pub trait AbstractAttributes {
    /// Add the provided `attrs`
    fn add_abstract_attributes<A: Into<Attribute>>(
        self,
        attrs: impl IntoIterator<Item = A>,
    ) -> Self;
}

/// Trait for adding `abstract` events to a [`Response`]
impl AbstractAttributes for Response {
    fn add_abstract_attributes<A: Into<Attribute>>(
        mut self,
        attrs: impl IntoIterator<Item = A>,
    ) -> Self {
        // Find the index of the first abstract event in the events vector
        let index = self.events.iter().position(|e| e.ty == ABSTRACT_EVENT_TYPE);

        if let Some(index) = index {
            // If an abstract event exists, replace it with a new event that has the additional attributes
            let event = self.events.remove(index);
            let new_event = event.add_attributes(attrs);
            self.events.insert(index, new_event);
        } else {
            // If an abstract event does not exist, create a new one with the additional attributes
            let new_event = Event::new(ABSTRACT_EVENT_TYPE).add_attributes(attrs);
            self.events.push(new_event);
        }

        self
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use speculoos::prelude::*;

    #[test]
    fn test_add_abstract_attributes_no_abstract_event() {
        let mut response = Response::default();

        assert_that!(response.events).has_length(0);

        response = response.add_abstract_attributes(vec![("key1", "value1")]);

        assert_that!(response.events).has_length(1);
        let event = response.events.first().unwrap();
        assert_that!(event.ty).is_equal_to(ABSTRACT_EVENT_TYPE.to_string());
        assert_that!(event.attributes).has_length(1);

        let attr = event.attributes.first().unwrap();
        assert_that!(attr.key).is_equal_to("key1".to_string());
        assert_that!(attr.value).is_equal_to("value1".to_string());
    }

    #[test]
    fn test_add_abstract_attributes_with_abstract_event() {
        const KEY_1: &str = "abstract_key1";
        const VALUE_1: &str = "abstract_value1";
        let mut response = Response::default()
            .add_event(Event::new(ABSTRACT_EVENT_TYPE).add_attributes(vec![(KEY_1, VALUE_1)]));

        assert_that!(response.events).has_length(1);
        let event = response.events.first().unwrap();
        assert_that!(event.ty).is_equal_to(ABSTRACT_EVENT_TYPE.to_string());
        assert_that!(event.attributes).has_length(1);

        // Add the attributes
        const KEY_2: &str = "abstract_key2";
        const VALUE_2: &str = "abstract_value2";
        response = response.add_abstract_attributes(vec![(KEY_2, VALUE_2)]);

        assert_that!(response.events).has_length(1);
        let event = response.events.first().unwrap();
        assert_that!(event.ty).is_equal_to(ABSTRACT_EVENT_TYPE.to_string());
        // should now be 2
        assert_that!(event.attributes).has_length(2);

        let attr1 = event.attributes.first().unwrap();
        assert_that!(attr1.key).is_equal_to(KEY_1.to_string());
        assert_that!(attr1.value).is_equal_to(VALUE_1.to_string());

        let attr2 = event.attributes.get(1).unwrap();
        assert_that!(attr2.key).is_equal_to(KEY_2.to_string());
        assert_that!(attr2.value).is_equal_to(VALUE_2.to_string());
    }
}
