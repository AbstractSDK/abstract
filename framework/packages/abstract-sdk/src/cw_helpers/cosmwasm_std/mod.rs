mod abstract_attributes;
mod wasm_query;

pub use abstract_attributes::AbstractAttributes;
pub use wasm_query::wasm_raw_query;
pub(crate) use wasm_query::ApiQuery;
