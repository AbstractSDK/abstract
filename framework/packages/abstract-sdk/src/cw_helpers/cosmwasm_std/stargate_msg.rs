use cosmos_sdk_proto::traits::Message;
use cosmwasm_std::to_binary;
use cosmwasm_std::CosmosMsg;
use cosmwasm_std::StdResult;
use serde::Serialize;

/// Shortcut helper as the construction of CosmosMsg::Stargate {...} can be quite verbose in contract code.
/// ```rust,norun
///         let msg = MsgCreateDenom {
///             sender: self.sender()?,
///             subdenom: self.subdenom.to_string(),
///         };
///
///         let msg = prost_stargate_msg(MsgCreateDenom::TYPE_URL, &msg)?;
/// ```
pub fn prost_stargate_msg(
    type_url: impl Into<String>,
    msg: impl Message,
) -> StdResult<cosmwasm_std::CosmosMsg> {
    Ok(cosmwasm_std::CosmosMsg::Stargate {
        type_url: type_url.into(),
        value: cosmwasm_std::Binary(msg.encode_to_vec()),
    })
}

#[cfg(test)]
mod test {
    use cosmos_sdk_proto::cosmos;

    use super::*;

    #[test]
    fn test_stargate_msg() {
        let msg = cosmos::feegrant::v1beta1::MsgRevokeAllowance {
            granter: "foo".to_owned(),
            grantee: "bar".to_owned(),
        };
        let _msg = prost_stargate_msg("/cosmos.feegrant.v1beta1.MsgGrantAllowance", &msg).unwrap();
    }
}
