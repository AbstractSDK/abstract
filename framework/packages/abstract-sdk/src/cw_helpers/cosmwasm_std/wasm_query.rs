use cosmwasm_std::{to_binary, QueryRequest, StdResult, WasmQuery};
use serde::Serialize;

/// Shortcut helper as the construction of QueryRequest::Wasm(WasmQuery::Smart {...}) can be quite verbose in contract code
pub fn wasm_smart_query<C>(
    contract_addr: impl Into<String>,
    msg: &impl Serialize,
) -> StdResult<QueryRequest<C>> {
    let query_msg = to_binary(msg)?;
    Ok(QueryRequest::Wasm(WasmQuery::Smart {
        contract_addr: contract_addr.into(),
        msg: query_msg,
    }))
}

/// Shortcut helper as the construction of QueryRequest::Wasm(WasmQuery::Raw {...}) can be quite verbose in contract code
pub fn wasm_raw_query<C>(
    contract_addr: impl Into<String>,
    key: &str,
) -> StdResult<QueryRequest<C>> {
    Ok(QueryRequest::Wasm(WasmQuery::Raw {
        contract_addr: contract_addr.into(),
        key: key.as_bytes().into(),
    }))
}

#[cfg(test)]
mod test {
    use super::*;
    use core::{app, app::BaseQueryMsg};
    use cosmwasm_std::Empty;

    #[test]
    fn test_wasm_smart_query() {
        let query_msg = app::QueryMsg::<Empty>::Base(BaseQueryMsg::BaseAdmin {});
        let query = wasm_smart_query::<Empty>("contract", &query_msg).unwrap();
        match query {
            QueryRequest::Wasm(WasmQuery::Smart { contract_addr, msg }) => {
                assert_eq!(contract_addr, "contract");
                assert_eq!(msg, to_binary(&query_msg).unwrap());
            }
            _ => panic!("Unexpected query"),
        }
    }

    #[test]
    fn test_wasm_raw_query() {
        let query = wasm_raw_query::<Empty>("contract", "key").unwrap();
        match query {
            QueryRequest::Wasm(WasmQuery::Raw { contract_addr, key }) => {
                assert_eq!(contract_addr, "contract");
                assert_eq!(key, cosmwasm_std::Binary::from("key".as_bytes()));
            }
            _ => panic!("Unexpected query"),
        }
    }
}
