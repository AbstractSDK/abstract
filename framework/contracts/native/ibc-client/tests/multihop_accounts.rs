use abstract_interface::Abstract;
use abstract_std::{
    ibc::polytone_callbacks::CallbackMessage,
    ibc_client::{ExecuteMsgFns as _, IbcClientCallback, QueryMsgFns},
    objects::{AccountId, TruncatedChainId},
    ABSTRACT_EVENT_TYPE,
};
use cosmwasm_std::{to_json_binary, StdResult, SubMsgResponse};
use cw_orch::{core::serde_json, mock::MockBech32, prelude::*, take_storage_snapshot};

type AResult = cw_orch::anyhow::Result<()>; // alias for Result<(), anyhow::Error>

#[test]
fn multihop_account_snapshot() -> AResult {
    let chain = MockBech32::new("mock");
    // Mock note, so it can take execute calls
    let note_code_id = chain
        .upload_custom(
            "note",
            Box::new(ContractWrapper::new(
                |_, _, _, _: serde_json::Value| StdResult::Ok(cosmwasm_std::Response::new()),
                |_, _, _, _: Empty| StdResult::Ok(cosmwasm_std::Response::new()),
                |_,
                 _,
                 _: cosmwasm_std::Empty|
                 -> Result<cosmwasm_std::Binary, cosmwasm_std::Never> {
                    unreachable!()
                },
            )),
        )?
        .uploaded_code_id()?;
    let note = chain
        .instantiate(note_code_id, &Empty {}, Some("note"), None, &[])?
        .instantiated_contract_address()?;

    // Make ibc-client trust our mock note for registering accounts
    let deployment = Abstract::new(chain.clone());
    deployment.ibc.client.upload()?;
    deployment
        .ibc
        .client
        .instantiate(&abstract_std::ibc_client::InstantiateMsg {}, None, &[])?;
    deployment.ibc.client.register_infrastructure(
        TruncatedChainId::from_chain_id("remoteone-1"),
        "host",
        note.clone(),
    )?;
    deployment
        .ibc
        .client
        .call_as(&note)
        .callback(CallbackMessage {
            initiator: deployment.ibc.client.address()?,
            initiator_msg: to_json_binary(&IbcClientCallback::WhoAmI {})?,
            result: abstract_std::ibc::polytone_callbacks::Callback::Execute(Ok(
                abstract_std::ibc::polytone_callbacks::ExecutionResponse {
                    executed_by: "host".to_owned(),
                    result: vec![],
                },
            )),
        })?;

    let multihop_account_id = AccountId::new(
        42,
        abstract_std::objects::AccountTrace::Remote(vec![
            TruncatedChainId::from_chain_id("remoteone-1"),
            TruncatedChainId::from_chain_id("remotetwo-1"),
        ]),
    )?;
    // register account
    deployment
        .ibc
        .client
        .call_as(&note)
        .callback(CallbackMessage {
            initiator: deployment.ibc.client.address()?,
            initiator_msg: to_json_binary(&IbcClientCallback::CreateAccount {
                account_id: multihop_account_id.clone(),
            })?,
            result: abstract_std::ibc::polytone_callbacks::Callback::Execute(Ok(
                abstract_std::ibc::polytone_callbacks::ExecutionResponse {
                    executed_by: "host".to_owned(),
                    #[allow(deprecated)]
                    result: vec![SubMsgResponse {
                        events: vec![cosmwasm_std::Event::new(ABSTRACT_EVENT_TYPE)
                            .add_attribute("account_address", "remote_account")],
                        data: None,
                        msg_responses: vec![],
                    }],
                },
            )),
        })?;
    let accounts = deployment.ibc.client.list_accounts(None, None)?;
    // Make sure we have in state exactly what we did put
    assert_eq!(
        accounts.accounts,
        vec![(
            multihop_account_id,
            TruncatedChainId::from_chain_id("remoteone-1"),
            "remote_account".to_owned()
        )]
    );

    take_storage_snapshot!(chain, "multihop_account");
    Ok(())
}
