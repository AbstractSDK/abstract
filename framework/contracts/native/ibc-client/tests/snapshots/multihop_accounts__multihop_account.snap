---
source: contracts/native/ibc-client/tests/multihop_accounts.rs
expression: all_storage
snapshot_kind: text
---
"abstract:ibc-client":
  - - "\u0000\u0002earemoteone"
    - "{\"polytone_note\":\"mock1mzdhwvvh22wrt07w59wxyd58822qavwkx5lcej7aqfkpqqlhaqfsetqc4t\",\"remote_abstract_host\":\"host\",\"remote_proxy\":\"host\"}"
  - - "\u0000\u0002ebmock1mzdhwvvh22wrt07w59wxyd58822qavwkx5lcej7aqfkpqqlhaqfsetqc4t"
    - "\"remoteone\""
  - - "\u0000\u0002ec\u0000\tremoteone\u0000\tremotetwo\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0004\u0000\u0000\u0000*remoteone"
    - "\"remote_account\""
  - - contract_info
    - "{\"contract\":\"abstract:ibc-client\",\"version\":\"0.25.0\"}"
  - - mod
    - "{\"module\":\"abstract:ibc-client\",\"version\":\"0.25.0\",\"dependencies\":[],\"metadata\":null}"
  - - ownership
    - "{\"owner\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\",\"pending_owner\":null,\"pending_expiry\":null}"
